src/Ormolu/Utils/Extensions.hs

```diff
@@ -0,0 +1,117 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+
+module Ormolu.Utils.Extensions
+  ( Extension (..),
+    getExtensionsFromCabalFile,
+    findCabalFile,
+    getCabalExtensionDynOptions,
+  )
+where
+
+import Control.Exception
+import Control.Monad.IO.Class
+import qualified Data.ByteString as B
+import Data.List (find)
+import Data.Map.Lazy (Map)
+import qualified Data.Map.Lazy as M
+import Data.Maybe (maybeToList)
+import qualified Distribution.ModuleName as ModuleName
+import Distribution.PackageDescription
+import Distribution.PackageDescription.Parsec
+import Distribution.Utils.Path
+import Language.Haskell.Extension
+import Ormolu.Config
+import System.Directory
+import System.FilePath
+import System.IO.Error (isDoesNotExistError)
+
+getExtensionsFromCabalFile ::
+  MonadIO m =>
+  -- | Path to cabal file
+  FilePath ->
+  -- | Map from Haskell source file paths (without any extensions)
+  -- to its default language extensions
+  m (Map FilePath [DynOption])
+getExtensionsFromCabalFile cabalFile = liftIO $ do
+  Just GenericPackageDescription {..} <-
```

Comment:
I think that rather than having a pattern match just on `Just` we should add a new `OrmoluException` specifically for Cabal files that fail to parse. We could assign exit code 8 to that.

Reply:
Right now, the functions in the `Ormolu` module don't call functions from `Ormolu.Utils.Extensions` at all. Do we want to add a variant to `OrmoluException` that can never be thrown there, or add a new exception type in `Ormolu.Utils.Extension`?

Reply:
I think we should add a new data constructor to `OrmoluException` and then benefit from the existing helpers such as `withPrettyOrmoluExceptions` (which wraps the invocations of `getExtensionsFromCabalFile` in the `Main` module).